src/geometry/PolySetBuilder.cc

```diff
@@ -0,0 +1,148 @@
+ *  software in the executable aside from CGAL.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include "PolySetBuilder.h"
+#include <PolySet.h>
+#include "Geometry.h"
+#include "cgalutils.h"
+
+
+#include <memory>
+
+#ifdef ENABLE_MANIFOLD
+#include "ManifoldGeometry.h"
+#endif
+
+
+#include "CGAL_Nef_polyhedron.h"
+#include "CGALHybridPolyhedron.h"
+
+PolySetBuilder::PolySetBuilder(int vertices_count, int indices_count, int dim, boost::tribool convex)
+{
+  ps= new PolySet(dim, convex);
+  if(vertices_count != 0) ps->vertices.reserve(vertices_count);
+  if(indices_count != 0) ps->indices.reserve(indices_count);
+}
+
+PolySetBuilder::PolySetBuilder(const Polygon2d pol)
+{
+  ps= new PolySet(2, unknown);
+  ps->polygon=std::move(pol);   
+  ps->dirty=false;
+}
+
+int PolySetBuilder::vertexIndex(const Vector3d &pt)
+{
+  return allVertices.lookup(pt);
+}
+
+void PolySetBuilder::append_poly(const std::vector<int> &inds)
+{
+  ps->indices.push_back(inds);        
+  ps->dirty = true;
+}
+
+void PolySetBuilder::append_geometry(const shared_ptr<const Geometry>& geom)
+{
+  if (const auto geomlist = dynamic_pointer_cast<const GeometryList>(geom)) {
+    for (const Geometry::GeometryItem& item : geomlist->getChildren()) {
+      append_geometry(item.second);
+    }
+  } else if (const auto N = dynamic_pointer_cast<const CGAL_Nef_polyhedron>(geom)) {
+    PolySet ps(3);
+    bool err = CGALUtils::createPolySetFromNefPolyhedron3(*(N->p3), ps);
+    if (err) {
+      LOG(message_group::Error, "Nef->PolySet failed");
+    } else {
+      append(&ps);
+    }
+  } else if (const auto ps = dynamic_pointer_cast<const PolySet>(geom)) {
+    append(ps.get());
+  } else if (const auto hybrid = dynamic_pointer_cast<const CGALHybridPolyhedron>(geom)) {
+    // TODO(ochafik): Implement append_geometry(Surface_mesh) instead of converting to PolySet
+    append(hybrid->toPolySet().get());
+#ifdef ENABLE_MANIFOLD
+  } else if (const auto mani = dynamic_pointer_cast<const ManifoldGeometry>(geom)) {
+   append(mani->toPolySet().get());
+#endif
+  } else if (dynamic_pointer_cast<const Polygon2d>(geom)) { // NOLINT(bugprone-branch-clone)
+    assert(false && "Unsupported file format");
+  } else { // NOLINT(bugprone-branch-clone)
+    assert(false && "Not implemented");
+  }
+
+}
+
+
+void PolySetBuilder::append_poly(const std::vector<Vector3d> &v)
+{
+  std::vector<int> inds;
+  for(const auto &pt: v)
+    inds.push_back(vertexIndex(pt));  
+  ps->indices.push_back(inds);        
+  ps->dirty = true;
```

Comment:
Isn't the convention that you generate a valid PolySet in the `result()` function, so setting the PolySet dirty should perhaps be done there?